### PR TITLE
feat(components): [Dialog & MessageBox] support draggable

### PR DESCRIPTION
**Pull request by kooriookami**
This PR is based on branch `feat/dialog-draggable`.
feat(dialog): add draggable function #2718

### CODE REVIEW

File: docs/en-US/component/dialog.md [hunk 1]
```diff
@@ -117,6 +127,7 @@ When using `modal` = false, please make sure that `append-to-body` was set to **
 | close-on-press-escape | whether the Dialog can be closed by pressing ESC                                                  | boolean                                          | —               | true    |
 | show-close            | whether to show a close button                                                                    | boolean                                          | —               | true    |
 | before-close          | callback before Dialog closes, and it will prevent Dialog from closing                            | function(done)，done is used to close the Dialog | —               | —       |
+| draggable             | whether can drag Dialog                                                                           | boolean                                          | —               | false   |
```
Comment (JeremyWuuuuu) [inline docs/en-US/component/dialog.md:130]:
```suggestion
| draggable | enable dragging feature for Dialog | boolean | — | false |
```

File: docs/en-US/component/message-box.md [hunk 2]
```diff
@@ -95,6 +95,16 @@ message-box/centered-content
 
 :::
 
+## Draggable
+
+MessageBox can be drag.
```
Comment (JeremyWuuuuu) [inline docs/en-US/component/message-box.md:100]:
```suggestion
MessageBox can be draggable
```

File: docs/en-US/component/message-box.md [hunk 3]
```diff
@@ -95,6 +95,16 @@ message-box/centered-content
...
+
+:::demo Setting `draggable` to `true` can drag MessageBox
```
Comment (JeremyWuuuuu) [inline docs/en-US/component/message-box.md:102]:
```suggestion
:::demo Setting `draggable` to `true` allows user to drag message box
```

File: docs/en-US/component/message-box.md [hunk 4]
```diff
@@ -147,5 +157,6 @@ The corresponding methods are: `ElMessageBox`, `ElMessageBox.alert`, `ElMessageB
...
 | center                       | whether to align the content in center                                                                                                   | boolean                                                                                                                                                                                                                         | —                                | false                                            |
+| draggable                    | whether can drag MessageBox                                                                                                              | boolean                                                                                                                                                                                                                         | —                                | false                                            |
```
Comment (JeremyWuuuuu) [inline docs/en-US/component/message-box.md:160]:
```suggestion
| draggable                    | whether MessageBox is draggable                                                                                                           | boolean                                                                                                                                                                                                                         | —                                | false                                            |
```

File: packages/components/dialog/src/use-dialog.ts [hunk 5]
```diff
@@ -159,6 +165,10 @@ export const useDialog = (
       rendered.value = true // enables lazy rendering
       open()
     }
+    if (!props.fullscreen && props.draggable) {
+      headerRef.value = targetRef.value.querySelector('.el-dialog__header')
+      useDraggable(targetRef, headerRef)
+    }
```
Comment (JeremyWuuuuu) [inline packages/components/dialog/src/use-dialog.ts:171]: Can we use `ref` on the template?

Comment (JeremyWuuuuu) [reply packages/components/dialog/src/use-dialog.ts:171]: For instance:

```vue
<div class="el-dialog__header" ref="headerRef">
</div>

```

File: packages/components/message-box/src/index.vue [hunk 6]
```diff
@@ -270,6 +281,8 @@ export default defineComponent({
       () => state.icon || TypeComponentsMap[state.type] || ''
     )
     const hasMessage = computed(() => !!state.message)
+    const rootRef = ref<ComponentPublicInstance>(null)
+    const headerRef = ref<ComponentPublicInstance>(null)
```
Comment (JeremyWuuuuu) [inline packages/components/message-box/src/index.vue:285]: **Question:**  I think this one should be `HTMLElement` per your usage on the template, right?

File: packages/components/message-box/src/index.vue [hunk 7]
```diff
@@ -316,6 +329,9 @@ export default defineComponent({
       if (props.closeOnHashChange) {
         on(window, 'hashchange', doClose)
       }
+      if (props.draggable) {
+        useDraggable(rootRef, headerRef)
+      }
```
Comment (JeremyWuuuuu) [inline packages/components/message-box/src/index.vue:334]: **Suggestion:** I think we should make this reactive, when the value was set to false, we should be able to turn this feature off.

Comment (JeremyWuuuuu) [reply packages/components/message-box/src/index.vue:334]: You can enable watching in this `onMounted` block, also I didn't see your destroy handler, would you mind to add that?

**Review by jw-foss — COMMENTED** [review]
I left a few comments for you to consider, please let me know if you have any questions

**Review by jw-foss — APPROVED** [review]

**Pull request merged**